src/main/scala/chisel3/experimental/inlinetest/InlineTest.scala

```diff
@@ -26,38 +26,32 @@ class TestParameters[M <: RawModule, R] private[inlinetest] (
   final def desiredTestModuleName = s"test_${dutName}_${testName}"
 }
 
-/** Contains traits that implement behavior common to generators for unit test testharness modules. */
-object TestHarness {
-  import chisel3.{Module => ChiselModule, RawModule => ChiselRawModule}
-
-  /** TestHarnesses for inline tests without clock and reset IOs should extend this. This
-    * trait sets the correct desiredName for the module, instantiates the DUT, and provides
-    * methods to elaborate the test.
-    *
-    *  @tparam M the type of the DUT module
-    *  @tparam R the type of the result returned by the test body
-    */
-  trait RawModule[M <: ChiselRawModule, R] extends Public { this: ChiselRawModule =>
-    def test: TestParameters[M, R]
-    override def desiredName = test.desiredTestModuleName
-    val dut = Instance(test.dutDefinition)
-    final def elaborateTest(): R = test.testBody(dut)
-  }
-
-  /** TestHarnesses for inline tests should extend this. This trait sets the correct desiredName for
-   *  the module, instantiates the DUT, and provides methods to elaborate the test. By default, the
+/** TestHarnesses for inline tests should extend this. This trait sets the correct desiredName for
+   *  the module, instantiates the DUT, and provides methods to generate the test. By default, the
    *  reset is synchronous, but this can be changed by overriding [[resetType]].
    *
    *  @tparam M the type of the DUT module
    *  @tparam R the type of the result returned by the test body
    */
-  trait Module[M <: ChiselRawModule, R] extends RawModule[M, R] { this: ChiselModule =>
-    override def resetType = test.resetType match {
-      case Some(rt @ Module.ResetType.Synchronous)  => rt
-      case Some(rt @ Module.ResetType.Asynchronous) => rt
-      case _                                        => Module.ResetType.Synchronous
-    }
+abstract class TestHarness[M <: RawModule, R](test: TestParameters[M, R]) extends Module with Public {
+  override def desiredName = test.desiredTestModuleName
+
+  override def resetType = test.resetType match {
+    case Some(rt @ Module.ResetType.Synchronous)  => rt
+    case Some(rt @ Module.ResetType.Asynchronous) => rt
+    case _                                        => Module.ResetType.Synchronous
   }
+
+  final val finish = IO(Output(Bool()))
+  final val success = IO(Output(Bool()))
+
+  // Handle the base case where a test has no result. In this case, we expect
+  // the test to end the simulation and signal pass/fail.
+  finish := false.B
+  success := true.B
+
+  protected val dut = Instance(test.dutDefinition)
+  protected val testResult = test.testBody(dut)
 }
```

Comment:
This needs to prevent further IO creation. There are two ways to do that:

1. Use `FixedIORawModule`:

``` scala
final class TestHarnessIO extends Bundle {
  final val clock = IO(Input(Clock()))
  final val reset = IO(Input(Bool())) // Or Reset?
  final val finish = IO(Output(Bool()))
  final val success = IO(Output(Bool()))
}

class TestHarness(params...) extends FixedIORawModule(new TestHarnessIO) {
  ...
}
```

2. Use the somewhat obscure `endIOCreation()` API ad the end of this module to prevent additional IO from being created. (`FixedIORawModule` already does this.)